router/sleeve.go

```diff
@@ -0,0 +1,1117 @@
+				fwd.crypto.EncDF, fwd.senderDF, fwd.maxPayload)
+
+		case special := <-specialChan:
+			err = fwd.handleSpecialFrame(special)
+
+		case _, ok := <-confirmedChan:
+			if !ok {
+				// specialChan is closed to indicate
+				// the forwarder is being closed
+				break loop
+			}
+
+			err = fwd.confirmed()
+
+		case <-timerChan(fwd.heartbeatTimer):
+			err = fwd.sendHeartbeat()
+
+		case <-timerChan(fwd.heartbeatTimeout):
+			err = fmt.Errorf("timed out waiting for UDP heartbeat")
+
+		case <-tickerChan(fwd.fragTestTicker):
+			err = fwd.sendFragTest()
+
+		case <-timerChan(fwd.pmtuTestTimeout):
+			err = fwd.handlePMTUTestFailure()
+		}
+	}
+
+	if fwd.heartbeatTimer != nil {
+		fwd.heartbeatTimer.Stop()
+	}
+	if fwd.heartbeatTimeout != nil {
+		fwd.heartbeatTimeout.Stop()
+	}
+	if fwd.fragTestTicker != nil {
+		fwd.fragTestTicker.Stop()
+	}
+	if fwd.pmtuTestTimeout != nil {
+		fwd.pmtuTestTimeout.Stop()
+	}
+
+	checkWarn(fwd.senderDF.close())
+
+	fwd.lock.RLock()
+	defer fwd.lock.RUnlock()
+	if fwd.listener != nil {
+		fwd.listener.Error(err)
+	}
+}
+
+func (fwd *sleeveForwarder) aggregateAndSend(frame aggregatorFrame,
+	aggChan <-chan aggregatorFrame, enc Encryptor, sender udpSender,
+	limit int) error {
+	// Give up after processing N frames, to avoid starving the
+	// other activities of the forwarder goroutine.
+	i := 0
+
+	for {
+		// Adding the first frame to an empty buffer
+		if !fits(frame, enc, limit) {
+			log.Println(fwd.logPrefix(), "Dropping too big frame during forwarding: frame len", len(frame.frame), ", limit ", limit)
+			return nil
+		}
+
+		for {
+			enc.AppendFrame(frame.src, frame.dst, frame.frame)
+			i++
+
+			gotOne := false
+			if i < 100 {
+				select {
+				case frame = <-aggChan:
+					gotOne = true
+				default:
+				}
+			}
+
+			if !gotOne {
+				return fwd.flushEncryptor(enc, sender)
+			}
+
+			// Accumulate frames until doing so would
+			// exceed the PMTU.  Even in the non-DF case,
+			// it doesn't seem worth adding a frame where
+			// that would lead to fragmentation,
+			// potentially delaying or risking other
+			// frames.
+			if !fits(frame, enc, fwd.maxPayload) {
+				break
+			}
+		}
+
+		if err := fwd.flushEncryptor(enc, sender); err != nil {
+			return err
+		}
+	}
+}
+
+func fits(frame aggregatorFrame, enc Encryptor, limit int) bool {
+	return enc.TotalLen()+enc.FrameOverhead()+len(frame.frame) <= limit
+}
+
+func (fwd *sleeveForwarder) flushEncryptor(enc Encryptor,
+	sender udpSender) error {
+	msg, err := enc.Bytes()
+	if err != nil {
+		return err
+	}
+
+	return fwd.processSendError(sender.send(msg, fwd.remoteAddr))
+}
+
+func (fwd *sleeveForwarder) sendSpecial(enc Encryptor, sender udpSender,
+	data []byte) error {
+	enc.AppendFrame(fwd.sleeve.localPeerBin, fwd.remotePeerBin, data)
+	return fwd.flushEncryptor(enc, sender)
+}
+
+func (fwd *sleeveForwarder) handleSpecialFrame(special specialFrame) error {
+	if special.sender == nil {
+		return fwd.handleControlMsg(special.frame)
```

Comment:
Yes, using a single goroutine leads to the large select, which was always a concern for me.  OTOH, using a single goroutine might also yield performance benefits to, due to improved locality etc.  My performance measurements suggest it's a wash.

The reason for using a single goroutine was simplicity.  Compared to the original code, there is less locking and inter-goroutine communication involved, because responsibilities that were dispersed between the Connection and the two forwarder goroutines are now confined to a single goroutine.